girder_volview/dicom.py

```diff
@@ -0,0 +1,138 @@
+import datetime
+import sys
+
+import pydicom
+import pydicom.valuerep
+import pydicom.multival
+import pydicom.sequence
+
+from girder import events
+from girder.models.item import Item
+from girder.models.file import File
+
+MAX_TAG_SIZE = 1024 * 128  # bytes
+
+
+def setupEventHandlers():
+    events.bind("model.file.save.after", "girder_volview", handleFileSave)
+
+
+def handleFileSave(event):
+    return addDicomTagsToItem(event.info)
+
+
+# Code modified from https://github.com/girder/girder/blob/master/plugins/dicom_viewer/girder_dicom_viewer/__init__.py
+def addDicomTagsToItem(file):
+    """
+    Add DICOM tags to Item metadata.
+    """
+    itemId = file["itemId"]
+    if not itemId:
+        return
+    dicomMetadata = _parseFile(file)
+    if dicomMetadata is None:
+        return  # not a dicom file
+    itemMeta = {"dicom": dicomMetadata}
+    item = Item().load(itemId, force=True)
+    Item().setMetadata(item, itemMeta)
+
+
+def _coerceValue(value):
+    # For binary data, see if it can be coerced further into utf8 data.  If
+    # not, mongo won't store it, so don't accept it here.
+    if isinstance(value, bytes):
+        if b"\x00" in value:
+            raise ValueError("Binary data with null")
+        try:
+            value.decode()
+            return bytes(value)
```

Comment:
I don't understand why you are calling `bytes` on a bytes instance.  I think we _want_ the decoded value so that it is a string (unicode) not bytes.  Otherwise, in the mongo json-ized storage we get the repr of the bytes object.  The current json looks like `"PatientName": "b'Value^Test'"`; we'd rather have `"PatientName": "Value^Test"`.

Reply:
I don't understand why ether =/  Was replicating the flow from the Dicom Viewer plugin.

Turns out PatientName is a  `pydicom.valuerep.PersonName` type and was not getting handled in the `if isinstance(value, bytes)` block.  PatientName went here:
```
    if isinstance(value, pydicom.valuerep.PersonName):
        return value.encode('utf-8') # returns type bytes =(
```
I changed that to 
```
    if isinstance(value, pydicom.valuerep.PersonName):
        return str(value)
```

Regarding the casting of a `bytes` to `bytes`: Casting up the inheretence highearchy (if that ever happended) does not change anything in my tests.  So I removed the cast and just return the value after the `value.decode()` check:
```
        try:
            value.decode()
            return value
        except UnicodeDecodeError:
            raise ValueError("Binary data that cannot be stored as utf-8")
```

Reply:
Now that this code deals with PatientName better, if a tag value really is binary, we are passing that information along with above/current code.  We could just return the decoded string value if you think that's better.  

Only DICOM file I have that has a binary value tag seems to be something propritary `"(0013, 1013)": "b'77779002'",`

[1.3.6.1.4.1.14519.5.2.1.7777.9002.276979399257180200489703215689.2551251219860422.zip](https://github.com/DigitalSlideArchive/girder_volview/files/14776508/1.3.6.1.4.1.14519.5.2.1.7777.9002.276979399257180200489703215689.2551251219860422.zip)

Reply:
None of my set has any remaining binary tags either.